src/tree/updater_gpu_hist.cu

```diff
@@ -139,49 +139,95 @@ __forceinline__ __device__ int BinarySearchRow(
  * device. Does not own underlying memory and may be trivially copied into
  * kernels.*/
 struct ELLPackMatrix {
+  /*! \brief How is the compressed data laid out? */
```

Comment:
I'm ok with specialization as optimization is never easy.  But I would like to have a very distinguishable and organized implementations for each specialization.  Polymorphism is one way to do it, another way may be by creating a dispatcher.